StunClient.cs

```diff
@@ -825,7 +826,7 @@ internal class StunClient
         /// <param name="socket">UDP socket to use.</param>
         /// <returns>Returns UDP netwrok info.</returns>
         /// <exception cref="Exception">Throws exception if unexpected error happens.</exception>
-        internal static StunResult Query(string host, int port, Socket socket)
+        internal static async Task<StunResult> QueryAsync(string host, int port, Socket socket)

```

Comment:
can we leave changes to the StunClient away from this PR? we can leave them for another PR

Reply:
I have to use the async overload of the `Query` method, otherwise I have to change the implementation of `ConnectAsync` to

```
static Task<UdtSocket> ConnectAsync(Socket socket, string remoteAddress, int remotePort, CancellationToken ct)
{
    try
    {
        GetExternalEndPointAsync(socket);

        var client = new UdtSocket(socket.AddressFamily, socket.SocketType);
        client.Bind(socket);

        Console.WriteLine("Trying to connect to {0}:{1}.  ", remoteAddress, remotePort);

        client.Connect(new IPEndPoint(IPAddress.Parse(remoteAddress), remotePort));

        Console.WriteLine("Connected successfully to {0}:{1}", remoteAddress, remotePort);

        if (ct.IsCancellationRequested)
        {
            socket.Close();
            ct.ThrowIfCancellationRequested();
        }

        return Task.FromResult(client);
    }
```

And all the tasks

```var taskList = remotePorts .Select((port, index) => ConnectAsync(Sockets[index], remoteAddress, port, cts.Token)).ToList();```

Will be executed synchronously.

Reply:
I understand that asynchronicity is better, but in this case it doesn't help because p2pcopy is a command-line tool anyway (so there's no need to send workload to the background while the UI runs) so there's not much difference here. Especially considering that the part that this PR needs to change is the connection between peers, not the connection to Stun servers (which worked fine, even if synchronously). We can make the connection to STUN asynchronous later in another PR.

Reply:
The purpose of this PR was to parallelize the connection to the client through different ports. By the way, I can remove all async codes and just try each port, one after another synchronously if you think there's not much difference between sync and async here.

Reply:
> The purpose of this PR was to parallelize the connection to the client through different ports.

Parallelize the connection to the peer, not to the STUN server.

> By the way, I can remove all async codes and just try each port, one after another synchronously if you think there's not much difference between sync and async here.

No, leave the connection to the peer asynchronous as you have already refactored it like that. But don't touch Stun-related code, it's all am asking.

Reply:
`UdtSocket` itself doesn't have any async operation and `STUN` server makes a transaction with the `TcpSocket` so I have to make this operation async to run tasks in the background.  As I posted the other version of `ConnectAsync` that returns `Task.FromResult(client);`, it will be executed synchronously and it doesn't help much. Another way is changing `ConnectAsync` to sync the version and running it with `Task.Run` in the background.

Reply:
I don't understand why you need to run any STUN-related stuff in the background, this was never asked for.

Reply:
(The 10 threads/tasks thing shouldn't start until after the connection with STUN servers has ended.)

Reply:
Could you please review my last commit?

Reply:
looking good! I think we're only missing a couple of nitpicks more